doc/decisions/ensure.md

```diff
@@ -0,0 +1,39 @@
+# Ensure
+
+## Problem
+
+Applications want to ensure that some functionality (=global plugin)
+is present in Elektra.
+
+## Constraints
+
+- keep notifications and highlevel API as-is
+
+## Assumptions
+
+## Considered Alternatives
+
+- integrate in kdbOpen
+
+## Decision
+
+Keep kdbEnsure but:
+
+- reduce for only global plugins, as the partial other functionality is confusing
+  and not needed
+- find solution that list plugin is not needed
+- only as implementation detail below libraries (e.g. like done for notification)
+
+## Rationale
+
+- allows interfaces that let the user do `kdbOpen`
```

Comment:
I clarified in f8d940d0da47d60c4d5f56d8f2740a3e5fc6d188

Yes, I also would have liked it in kdbOpen more but it seems unrealistic to me that this gets implemented as it needs heavy changes in the notification API. And there are also advantages to `kdbEnsure`, it is more flexible :wink:

Reply:
I really don't like the way `kdbEnsure` works right now. We set up the default case in `kdbOpen` only to potentially redo everything in `kdbEnsure`. Maybe the solution is an API like this:

```c
KDB * kdbOpen (Key * parent);
int kdbConfigure (KDB * handle, KeySet * contract, Key * parentKey);

KDB * kdbOpenDefault (Key * parent);
```

The new `kdbOpen` only does the absolute minimum work, in particular it doesn't set up any global plugins. If you use `kdbOpen` you _must_ call `kdbConfigure` otherwise `kdbGet` will fail. `kdbConfigure` configures global plugins (basically just a renamed `kdbEnsure`). Lastly, `kdbOpenDefault` does more or less what the old `kdbOpen` does. It sets up the default case and you can call `kdbGet` immediately. _But_ you _cannot_ call `kdbConfigure` after `kdbOpenDefault`.

(The names are just suggestions)

---

The high-level API is a bad example for why we should keep `kdbEnsure`. It takes the contract in `elektraOpen`, so it really doesn't matter.

---

The notifications API only works because of the `list` plugin. Without the `list` plugin we need considerable changes. In particular, `zeromqrecv` and `dbusrecv` must be mounted _before_ `elektraIoSetBinding` is called. That means you need to do:

```c
KDB * kdb = kdbOpen (key);
kdbEnsure (kdb, contract, key); // contract must contain `zeromqrecv` or `dbusrecv`
elektraIoSetBinding (kdb, binding);
elektraNotificationOpen (kdb); // calls kdbEnsure again to mount `internalnotification`
```

It would be significantly simpler and much less error-prone, if `kdbOpen` took the contract. Then we could do:

```c
KDB * kdb = kdbOpen (key, contract); // contract contains notification config
elektraIoSetBinding (kdb, binding);
elektraNotificationOpen (kdb); // fails if kdb is incorrectly configured
```

`kdbOpen` would mount `internalnotification` and `zeromqrecv` or `dbusrecv`, depending on the notification config in the `contract`.

The alternative would be to move the ["deferred call"](https://github.com/ElektraInitiative/libelektra/blob/f3ab8d68be71a6bbd46b6d478dd8cc6683dbfecc/doc/decisions/deferred_plugin_calls.md) functionality from `list` to the core and thereby allowing `elektraIoSetBinding` to be called before `zeromqrecv` or `dbusrecv` are mounted. I don't think this would be a good idea, it is a lot more complex than merging `kdbOpen` and `kdbEnsure`.

---

The `gopts` "wrapper" also can't be seen as an argument for a separate `kdbEnsure`. Because what we want actually is **not** a wrapper. We want something like `elektraSetIoBinding` or `elektraNotificationOpen` that is called after `gopts` is mounted and before `kdbGet`. What we want is actually this:

```c
KDB * kdb;
/* set up kdb such that `gopts` is mounted */
elektraGOptsSetArgs (kdb, argc, argv);

kdbGet (kdb, ks, parent);
```

`elektraGOptsSetArgs` finds the `Plugin *` for `gopts` in `kdb` and stores `argc` and `argv` in there, so that they can be used in `kdbGet`.

---

>  there are also advantages to kdbEnsure, it is more flexible

More flexible also means more error prone. (at least in this case)

Reply:
Ok, convinced, I turned around the whole decision: 3e79ed7a1747765ca6a427b5fa506878aa6da995